auto_nag/db.py

```diff
@@ -0,0 +1,422 @@
+
+        e = Tool(name)
+        session.add(e)
+        session.commit()
+
+        return e
+
+    def __repr__(self):
+        return self.name
+
+    def __str__(self):
+        return self.__repr__()
+
+
+class BugChange(Base):
+    __tablename__ = 'bugchanges'
+
+    id = Column(Integer, primary_key=True, autoincrement=True)
+    tool_id = Column(Integer, ForeignKey('tools.id', ondelete='CASCADE'))
+    extra_id = Column(Integer, ForeignKey('extras.id', ondelete='CASCADE'))
+    date = Column(Integer)
+    bugid = Column(Integer)
+
+    def __init__(self, tool, date, bugid, extra):
+        self.tool = Tool.get(tool)
+        self.date = get_ts(date)
+        self.bugid = int(bugid)
+        self.extra = Extra.get(extra)
+
+    def get_date(self):
+        return lmdutils.get_date_from_timestamp(self.date)
+
+    @staticmethod
+    def add(tool, bugid, ts=lmdutils.get_timestamp('now'), extra=''):
+        with FileLock(lock_path):
+            session.add(BugChange(tool, ts, bugid, extra))
+            session.commit()
+
+    @staticmethod
+    def get(name=None, start_date=None, end_date=None):
+        with FileLock(lock_path):
+            start_date = get_ts(start_date, default=0)
+            end_date = get_ts(start_date, default='now')
+            if name:
+                rs = (
+                    session.query(BugChange)
+                    .join(BugChange.tool)
+                    .filter(
+                        Tool.name == name,
+                        BugChange.date >= start_date,
+                        BugChange.date < end_date,
+                    )
+                )
+            else:
+                rs = session.query(BugChange).filter(
+                    BugChange.date >= start_date, BugChange.date < end_date
+                )
+
+            return [r for r in rs]
+
+    @staticmethod
+    def has_already_nagged(bugids, name=None, start_date=None, end_date=None):
+        with FileLock(lock_path):
+            data = {int(bugid): False for bugid in bugids}
+            bugids = list(data.keys())
+            start_date = get_ts(start_date, default=0)
+            end_date = get_ts(start_date, default='now')
+            if name:
+                res = (
+                    session.query(BugChange.bugid)
+                    .join(BugChange.tool)
+                    .filter(
+                        Tool.name == name,
+                        BugChange.bugid.in_(bugids),
+                        BugChange.date >= start_date,
+                        BugChange.date < end_date,
+                    )
+                )
+            else:
+                res = session.query(BugChange.bugid).filter(
+                    BugChange.bugid.in_(bugids),
+                    BugChange.date >= start_date,
+                    BugChange.date < end_date,
+                )
+
+            for r in res:
+                data[r.bugid] = True
+            return data
+
+    @staticmethod
+    def dump(path=''):
```

Comment:
A db structure change will make that really complex

Reply:
If you switch to alembic, you can make data migrations too (with forward/rollback)